### PR TITLE
[Week06] migrate to gymnasium

### PR DESCRIPTION
**Pull request by laktionov**
- Support new interfaces in all files and notebooks
- Change `done` to `terminated or truncated` to iterate over env
- Change `done` to `terminated` to check whether bootstrap or not when game ends
- Update video recording code
- Replace custom to_one_hot with F.one_hot in reinforce_pytorch.ipynb
- Update CartPole to v1 in reinforce_pytorch.ipynb
- Remove redundant imports from py files
- In atari_wrappers.py add `SwapImageAxes` wrapper to permute image axes and normalize pixels, replace `TFSummaries ` with `TensorboardSummaries` to log metrics in tensorboard, add auxiliary class `thunk_` (similar to [this](https://github.com/cydoroga/ftiad_rl_2021/blob/master/hw4/atari_wrappers.py) )

### CODE REVIEW

File: week06_policy_based/atari_wrappers.py [hunk 1]
```diff
@@ -274,53 +320,24 @@ def reset(self, **kwargs):
...
+    def __call__(self):
+        return nature_dqn_env(
+            self.env_id,
+            summaries=False,
+            clip_reward=False,
+            **self.kwargs,
+        )
 
 
-class NumpySummaries(SummariesBase):
```
Comment (q0o0p) [inline week06_policy_based/atari_wrappers.py:297]: I don't think we should remove `NumpySummaries`.
Many students still prefer to use it.
I think we should keep the ability to choose the type of summaries via argument passed to `nature_dqn_env()`.

Comment (laktionov) [reply week06_policy_based/atari_wrappers.py:297]: Thank you for your comment!
I guess the students still use it due to the lack of alternatives since `TFSummaries` has been deprecated.

Anyway, I've returned `SummariesBase` and `NumpySummaries` and replaced `TFSummaries` with `TensorboardSummaries`.

Comment (anna-vb) [reply week06_policy_based/atari_wrappers.py:297]: Thank you!

File: week06_policy_based/reinforce_pytorch.ipynb [hunk 2]
```diff
@@ -361,7 +336,7 @@
     "\n",
     "    # cast everything into torch tensors\n",
     "    states = torch.tensor(states, dtype=torch.float32)\n",
-    "    actions = torch.tensor(actions, dtype=torch.int32)\n",
+    "    actions = torch.tensor(actions, dtype=torch.long)\n",
```
Comment (dniku) [inline week06_policy_based/reinforce_pytorch.ipynb:339]: I think `long` is less explicit than `int32` (I don't remember off the top of my head if it's `int32` or `int64`) -- maybe let's keep this explicit?

**Review by dniku — COMMENTED** [review]
Have you tested this end-to-end?

File: week06_policy_based/reinforce_pytorch.ipynb (hunk 2 above)
Comment (laktionov) [reply week06_policy_based/reinforce_pytorch.ipynb]: Sure, now it's explicit

File: week06_policy_based/atari_wrappers.py [hunk 3]
```diff
@@ -146,126 +151,159 @@ def observation(self, observation):
...
+        old_shape = self.observation_space.shape
+        self.observation_space = Box(
+            low=0.0,
+            high=1.0,
+            shape=(old_shape[-1], old_shape[1], old_shape[0]),
+            dtype=np.float32,
+        )
+
+    def observation(self, observation):
+        return np.swapaxes(observation, 2, 0).astype(np.float32) / 255.0
```
[the remark this answers is not in the document]
Comment (laktionov) [reply week06_policy_based/atari_wrappers.py]: Now it's fixed.
I also made ImagePreprocessing consistent with the current axis' order.

**Comment by laktionov** [comment]
> Have you tested this end-to-end?

Yes, I've run my version of the solved assignment in Colab

File: week06_policy_based/atari_wrappers.py [hunk 4]
```diff
@@ -108,31 +108,33 @@ def reset(self, **kwargs):
...
+                shape=(height, width, *self.observation_space.shape[2:]),
+                dtype=dtype,
             )
 
     def observation(self, observation):
         """Performs image preprocessing."""
         if self.grayscale:
             observation = cv2.cvtColor(observation, cv2.COLOR_RGB2GRAY)
-        observation = cv2.resize(observation, (self.width, self.height), cv2.INTER_AREA)
+        observation = cv2.resize(observation, (self.height, self.width), cv2.INTER_AREA)
```
Comment (dniku) [inline week06_policy_based/atari_wrappers.py:137]: I think this is wrong -- the correct order is [(width, height)](https://docs.opencv.org/4.x/da/d6e/tutorial_py_geometric_transformations.html). Confusing, I know :(

(To make things more confusing, torchvision [uses (height, width)](https://pytorch.org/vision/main/generated/torchvision.transforms.functional.resize.html) by default.)

Comment (laktionov) [reply week06_policy_based/atari_wrappers.py]: A little bit misleading :)
Now it's fixed.

**Comment by dniku** [comment]
Thanks!
